### ISSUE TITLE
load data from drive: tfsim.samplers.TFDatasetMultiShotMemorySampler

### ISSUE THREAD

**Issue by BakhtawarRehman** [tensorflow/similarity#281]
Hello,

I am trying to load data from disk using `tfsim.samplers.TFDatasetMultiShotMemorySampler`. However, i am facing problems. Any help'd be hiighly appreciated.

Below is my approach to find the solution to the problem:

- The dataset consists of 36 classes. It is divided into 36 folders with variable number of images in each.
-  class1 / image1, image2 ......;   class2 / image1, image2 ...... ... .. class36/ image1, image2..... and so on

```
num_known_classes = 36
class_list = random.sample(population = range(36), k=num_known_classes)
classes_per_batch = 36

examples_per_class_per_batch = 8

train_ds = tfsim.samplers.TFDatasetMultiShotMemorySampler(
    data_dir, 
    classes_per_batch = min(classes_per_batch, num_known_classes),
    splits="train",
    steps_per_epoch=4000,
    examples_per_class_per_batch=examples_per_class_per_batch,
    class_list=class_list
)


print("\n" + " Create Validation Data ".center(34, "#"))
val_ds = tfsim.samplers.TFDatasetMultiShotMemorySampler(
    data_dir,
    classes_per_batch=classes_per_batch,
    splits='test',
    total_examples_per_class=100,
    examples_per_class_per_batch=2,
)

```

Please Note: I can easily load the data using:

`train_ds = tf.keras.utils.image_dataset_from_directory(
  data_dir,
  validation_split=0.2,
  subset="training",
  seed=123,
  image_size=(img_height, img_width),
  batch_size=batch_size)`

**Comment by owenvallis** [comment]
Hi,

I think the main issue here is that the [TFDatasetMultiShotMemorySampler](https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/samplers/tfdataset_samplers.py) is meant to load from one of the datasets TensorFlow catalog dataset [page](https://www.tensorflow.org/datasets/catalog/overview).

If you already have a local copy of your images and they can all fit into memory at the same time, then I would recommend using the [MultiShotMemorySampler](https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/samplers/memory_samplers.py#L31).

However, if you have more images than you can hold in memory, you'll need to use the [TFRecordDatasetSampler](https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/samplers/tfrecords_samplers.py#L22). This will also require you to save your images into recordio files. The size of each file should be a multiple of your `examples_per_class_per_batch` and should contain contiguous blocks of the same class, e.g., examples 0-7 are class 1 and examples 8-15 are class 2, this would represent a single recordio file. The reason you need this is that the only efficient way to sample and construct the batches on the fly is to use the ds.interleave function.

I have an [example notebook](https://github.com/tensorflow/similarity/blob/kaggle/examples/supervised/kaggle_keras_tuner.ipynb) in the Kaggle Colab branch. I haven't worked on that Colab in a long time though, so it may need a few tweaks to work, but should be pretty straight forward.

Let me know if you run into any issues.

**Comment by BakhtawarRehman** [comment]
Hi Owen, Many thanks for the guidance. 

I am a bit confuse about this argument of MultiShotMemorySampler: class_list.
According to description given on github 

> class_list: Filter the list of examples to only keep those who belong to the supplied class list.

it is difficult to understand if it is for whole dataset or for each batch only.

i mean this should be different for each batch, no? to my understanding: each batch gets different classes (based on our selection ofcourse) and class_list is there to define if we want to use examples of the 'selected classes Vs. everything', ?
